### ISSUE TITLE
rework dep parsing support to be more modular

### ISSUE THREAD

**Issue by radhermit** [pkgcraft/pkgcraft#214]
This would probably be best using a parser combinator library like https://crates.io/crates/winnow.

**Comment by idealseal** [comment]
I've written the code to parse `<category>/<name>-<version>` with `winnow` for now, but I could look into parsing more advanced specs like slots, etc (which I have not looked at too closely for now). Would you like me to contribute that code?

**Comment by radhermit** [comment]
Sure, feel free to make a PR with it if you like. What I'm mainly looking for is better modularity (e.g. being able to use it in the restriction parser without having excess code duplication), custom error support, passing parser tests (the target data currently lives in TOML files in the testdata repo -- https://github.com/pkgcraft/pkgcraft-testdata/tree/db1514e476c981a717525f7de204cc241f721971/toml), and fairly similar (or exceeding) benchmarks to the current parser which can be run via `cargo criterion`.

I also have had a reworked parser branch locally for a while (over a year by now), but need more inspiration to clean things up and incorporate it into the API and restriction support in an acceptable fashion. :smile:

**Comment by idealseal** [comment]
Yeah the testing is also a little hard to work on. I tried to add snapshot tests with [`insta`](https://docs.rs/insta), but getting the `cargo-insta` tool to update the snapshot is not that easy.

Would you consider moving that code into a separate crate? I would imagine that an interface of the PMS spec in the type system would be a great unit of code that could be reused by other crates as well. 

Also, I haven't found time yet to get further with thing like version operators and slots etc sadly.

**Comment by radhermit** [comment]
> Yeah the testing is also a little hard to work on. I tried to add snapshot tests with [`insta`](https://docs.rs/insta), but getting the `cargo-insta` tool to update the snapshot is not that easy.

That looks like an interesting tool for (de)serialize and CLI tests at least. Feels a bit complex to me though.

> BTW I read somewhere that you can write development dependency only crates that link against test binaries. If the code behind the `test` feature is moved there, all tests could link against that crate and the `pkgcraft[test]` etc feature dependencies would go away.

But then the test crates would have to be published as well. Personally I don't find running tests that laborious at the moment, all it takes running something like `cargo nextest run --features test --tests -p pkgcraft`.

Also, the shared test data exists in a separate repo (and pulled in as a submodule) because many of the bindings use it as well for parsing and repo related tests. Moving it to a rust-centric format would defeat the purpose.

> I would imagine that an interface of the PMS spec for the rust type system would be a great unit of code that could be reused by other crates as well, maybe it should be put in a separate crate?

I'm not sure what exactly you're talking about here. Are you saying you'd like dep parsing and related type objects in a separate crate? I've considered such things but it's hard to split up since the parsing will continue to be dependent on the EAPI and the eapi module pulls in a lot of the rest of pkgcraft due to registering command builtins per EAPI. In short, it would take a lot of work under the current design that I'd rather spend doing other things.

But if it was ever done, I could probably see crates like pkgcraft-core (eapi, dep, shell), pkgcraft-repo (repo and pkg), pkgcraft-restrict (restriction types), and some other misc crates like pkgcraft-cli (cli-related tooling like arg parsing). Just typing that out makes it sound like a lot of work. :laughing:

**Comment by idealseal** [comment]
So I have implemented CPV and CPN parsers with `winnow` and thrown in some benchmarks. The `winnow` implementation is focused on providing good error messages at the moment. I will test the new implementation with the already existing benchmarks soon.

### Winnow

```
big                     time:   [940.66 ns 941.44 ns 942.22 ns]
                        change: [-1.4458% -1.3264% -1.2071%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  5 (5.00%) high mild
  2 (2.00%) high severe

small                   time:   [290.60 ns 290.98 ns 291.35 ns]
                        change: [-8.1886% -7.9768% -7.7707%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

     Running benches/version.rs (target/release/deps/version-7843160b47d0349b)
Gnuplot not found, using plotters backend
big_version             time:   [271.46 ns 271.68 ns 271.92 ns]
                        change: [+3.7286% +3.8717% +4.0244%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe

small_version           time:   [93.969 ns 94.238 ns 94.498 ns]
                        change: [-6.9702% -6.4818% -6.0133%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

letter                  time:   [78.886 ns 78.975 ns 79.078 ns]
                        change: [-1.9704% -1.7463% -1.5102%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe

revision                time:   [73.193 ns 73.460 ns 73.748 ns]
                        change: [+0.3299% +0.7228% +1.1402%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) high mild
  3 (3.00%) high severe

suffix                  time:   [114.58 ns 115.48 ns 116.40 ns]
                        change: [+2.9426% +3.8840% +4.8936%] (p = 0.00 < 0.05)
                        Performance has regressed.
```

### peg

```
big                     time:   [802.97 ns 803.98 ns 805.32 ns]
                        change: [+0.3627% +0.6594% +1.0589%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 10 outliers among 100 measurements (10.00%)
  3 (3.00%) high mild
  7 (7.00%) high severe

small                   time:   [227.42 ns 227.55 ns 227.68 ns]
                        change: [-0.9301% -0.8210% -0.7195%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
  2 (2.00%) high severe

     Running benches/version_cmp.rs (target/release/deps/version_cmp-325961bbac2a71a5)
Gnuplot not found, using plotters backend
big_version             time:   [314.70 ns 315.07 ns 315.47 ns]
                        change: [-1.9988% -0.0703% +1.1058%] (p = 0.95 > 0.05)
                        No change in performance detected.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

small_version           time:   [87.033 ns 87.088 ns 87.147 ns]
                        change: [+0.5284% +0.6426% +0.7458%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

letter                  time:   [59.148 ns 59.223 ns 59.297 ns]
                        change: [+0.2074% +0.3491% +0.4948%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild

revision                time:   [70.702 ns 70.772 ns 70.843 ns]
                        change: [+0.8193% +0.9405% +1.0834%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
  1 (1.00%) high severe

suffix                  time:   [95.793 ns 95.866 ns 95.942 ns]
                        change: [+0.9970% +1.1455% +1.2815%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
```

### Benchmarked strings

#### CPV

|benchmark | input
|-|-
|`big` |`very-long_and+complicated/rust-1-av3-3458345.7864536.72634535z_alpha_beta8734_alpha_pre7345_rc005_p20250101-r0783653`
|`small`| `cat/pkg-1.1.1`


#### Version

| benchmark | input
|-|-
|`big_version` |`3458345.7864536.72634535z_alpha_beta8734_alpha_pre7345_rc005_p20250101-r0783653`
|`small_version`|  `1.1.1`
|`letter`  |`1a`
|`revision`| `1-r234`
 `suffix`| `1_alpha6_p5`

**Comment by radhermit** [comment]
> So I have implemented CPV and CPN parsers with `winnow` and thrown in some benchmarks. The `winnow` implementation is focused on providing good error messages at the moment. I will test the new implementation with the already existing benchmarks soon.

Looks promising... if you want feedback on the code at this point feel free to open a draft PR.
